src/crimson/os/seastore/seastore_types.h

```diff
@@ -1005,41 +1010,541 @@ constexpr journal_seq_t JOURNAL_SEQ_MAX{
 // JOURNAL_SEQ_NULL == JOURNAL_SEQ_MAX == journal_seq_t{}
 constexpr journal_seq_t JOURNAL_SEQ_NULL = JOURNAL_SEQ_MAX;
 
-// logical addr, see LBAManager, TransactionManager
-using laddr_t = uint64_t;
-constexpr laddr_t L_ADDR_MIN = std::numeric_limits<laddr_t>::min();
-constexpr laddr_t L_ADDR_MAX = std::numeric_limits<laddr_t>::max();
-constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
-constexpr laddr_t L_ADDR_ROOT = L_ADDR_MAX - 1;
-constexpr laddr_t L_ADDR_LBAT = L_ADDR_MAX - 2;
+// logical offset, see LBAManager, TransactionManager
+using extent_len_t = uint32_t;
+constexpr extent_len_t EXTENT_LEN_MAX =
+  std::numeric_limits<extent_len_t>::max();
+
+using extent_len_le_t = ceph_le32;
+inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+  return ceph_le32(len);
+}
+
+namespace details {
+// some concept required by laddr_t
+
+template <typename L, typename R = L>
+concept is_three_way_comparable = requires(const L &l, const R &r) { l <=> r; };
+
+template <typename T>
+concept is_convertible_to_ostream = std::is_convertible_v<T, std::ostream &>;
+
+template <typename T>
+concept is_streamable = requires(std::ostream &os, T value) {
+  { os << value } -> is_convertible_to_ostream;
+};
+
+} // namespace details
+
+/**
+ * laddr_t
+ *
+ * laddr_t uses uint128_t to represent the logical address internally.
+ * The layout of laddr:
+ *
+ * [pool:8][shard:8][crush:32][random:31][metadata:1][recover:1][local_clone_id:32][offset:15]
```

Comment:
Change to [is_metadata:1][random:31] for larger possible laddr space for object data/metadata.

Reply:
Probably an obvious question -- why place metadata, recover, local_clone_id after random?

Reply:
The ```random``` field is intended to distinguish between objects of different (pool, shard, crush, clone), right?  We can't actually use it for both that and reserved, I think?  Perhaps those should be split?  metadata makes sense -- we might want to use different placement policy with anything with that bit set.  I'm less sure about ```recover```, though.  There aren't currently any cases I can think of where we might want to have more than one temp object for a particular object, but I'm not sure we want to make that impossible.

For temp objects, it'll still be possible to enumerate temp objects, right?

Reply:
> edit: ah, to distinguish between metadata and clones of the same object, right? metadata is more significant than local_clone_id so that data for different clones of the same object will sort together?

Yes, seems it's better to seperate them into different address space.


> The `random` field is intended to distinguish between objects of different (pool, shard, crush, clone), right? We can't actually use it for both that and reserved, I think?

You're right, comments here is wrong.

> Perhaps those should be split?

Reserved bits seem reasonable, but it's not easy to determine where those bits should place, eg. reserve for gloabl bits or object local bits.


> I'm less sure about `recover`, though. There aren't currently any cases I can think of where we might want to have more than one temp object for a particular object, but I'm not sure we want to make that impossible.
> 
> For temp objects, it'll still be possible to enumerate temp objects, right?

This bit is introduced due to static layout, please see the long documentation blow, currently it's not used.

Reply:
> We can't actually use it for both that and reserved

My understanding is the [random:31] portion can shrink anytime to allow some bits to serve other purposes in the future.

> For temp objects, it'll still be possible to enumerate temp objects, right?

The scope of such enumeration is a PG, right?

My understanding is if we want to enumerate/list temp objects efficiently within a PG, they might need to be arranged together there.

Reply:
> For temp objects, it'll still be possible to enumerate temp objects, right?

I think the 128-bit lba key implementation doesn't affect objects enumeration, as we always to first list onodes from the onode tree and then access the objects' data/omaps in the enumeration, and no matter how the extents' laddr arranges, we can always get the right laddrs through onodes. Am I right?